ksqldb-engine/src/main/java/io/confluent/ksql/engine/EngineExecutor.java

```diff
@@ -210,13 +209,16 @@ PullQueryResult executeTablePullQuery(
       }
       return result;
     } catch (final Exception e) {
-      final PullSourceType finalSourceType = sourceType;
-      final PullPhysicalPlanType finalPlanType = planType;
-      final RoutingNodeType finalRoutingNodeType = routingNodeType;
-      pullQueryMetrics.ifPresent(metrics -> metrics.recordErrorRate(1,
-          Optional.ofNullable(finalSourceType).orElse(PullSourceType.UNKNOWN),
-          Optional.ofNullable(finalPlanType).orElse(PullPhysicalPlanType.UNKNOWN),
-          Optional.ofNullable(finalRoutingNodeType).orElse(RoutingNodeType.UNKNOWN)));
+      if (plan == null) {
+        pullQueryMetrics.ifPresent(m -> m.recordErrorRateForNoResult(1));
+      } else {
+        final PullPhysicalPlan physicalPlan = plan;
```

Comment:
Hmm why we need to re-assign to a different variable?

Reply:
Yeah, it's a pain. It's because you can only reference final variables from inside a callback. The reason we have a non-final variable involved is that we want to reference the plan in the `catch` block _if_ the plan happens to have gotten built.

I started to refactor the method so this wouldn't be necessary, but it's not trivial, and believe it or not, I showed some restraint.